### PR TITLE
[SPARK-31316][SQL] SQLQueryTestSuite: Display the total generate time for generated java code.

### PR DESCRIPTION
**Pull request by beliefer**
### What changes were proposed in this pull request?
After my investigation, SQLQueryTestSuite spent a lot of time to generate java code.
This PR is related to https://github.com/apache/spark/pull/28081.
https://github.com/apache/spark/pull/28081 used to display compile time, but this PR used to display generate time.
This PR will add the following to `SQLQueryTestSuite`'s output.
```
=== Metrics of Whole Codegen ===

Total generate time: 82.640913862 seconds

Total compile time: 98.649663572 seconds
```

### Why are the changes needed?
Display the total generate time for generated java code.

### Does this PR introduce any user-facing change?
'No'.

### How was this patch tested?
Jenkins test.

### CODE REVIEW

**Comment by beliefer** [comment]
cc @cloud-fan @dongjoon-hyun

File: sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala [hunk 1]
```diff
@@ -567,6 +567,26 @@ object WholeStageCodegenExec {
   def isTooManyFields(conf: SQLConf, dataType: DataType): Boolean = {
     numOfNestedFields(dataType) > conf.wholeStageMaxNumFields
   }
+
+  // The whole codegen generates java code on the driver side and sends it to the Executor side
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala:571]: `whole codegen` -> `whole stage codegen`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala [hunk 2]
```diff
@@ -567,6 +567,26 @@ object WholeStageCodegenExec {
...
+  // for execution after compilation. The whole codegen can bring significant performance
+  // improvements in large data and distributed environments. However, in the test environment,
+  // due to the small amount of data, the time to generate Java code takes up a major part of the
+  // entire runtime. So we summarize the total generation time and output it to the execution log
+  // for easy analysis and view.
+  private val _generateJavaTime = new LongAccumulator
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala:577]: why do we need accumulator if it's a pure driver side thing?

Comment (beliefer) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala:577]: I just followed #28081

File: sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala (hunk 1 above)
Comment (beliefer) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala:571]: OK

File: sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala (hunk 2 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala:577]: #28081 is different as we compile the code at the executor side. Yea let's use `AtomicLong`

Comment (beliefer) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala:577]: @cloud-fan I learned the use of `LongAccumulator` for statistics in a distributed environment.Thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala [hunk 3]
```diff
@@ -567,6 +567,26 @@ object WholeStageCodegenExec {
   def isTooManyFields(conf: SQLConf, dataType: DataType): Boolean = {
     numOfNestedFields(dataType) > conf.wholeStageMaxNumFields
   }
+
+  // The whole stage codegen generates java code on the driver side and sends it to the Executor
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala:571]: `Java code`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala [hunk 4]
```diff
@@ -39,7 +39,7 @@ import org.apache.spark.sql.execution.metric.SQLMetrics
 import org.apache.spark.sql.internal.SQLConf
 import org.apache.spark.sql.types._
 import org.apache.spark.sql.vectorized.ColumnarBatch
-import org.apache.spark.util.Utils
+import org.apache.spark.util.{LongAccumulator, Utils}
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala:42]: unnecessary change

File: sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala [hunk 5]
```diff
@@ -567,6 +567,26 @@ object WholeStageCodegenExec {
   def isTooManyFields(conf: SQLConf, dataType: DataType): Boolean = {
     numOfNestedFields(dataType) > conf.wholeStageMaxNumFields
   }
+
+  // The whole stage codegen generates java code on the driver side and sends it to the Executor
+  // side for execution after compilation. The whole codegen can bring significant performance
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala:572]: `... sends it to the Executors for compilation and execution.`

We also compile code at the driver side, but it's only for checking so that we can fail earlier. The result is discarded.

Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala:572]: whole stage codegen

File: sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala [hunk 6]
```diff
@@ -567,6 +567,26 @@ object WholeStageCodegenExec {
...
+  // improvements in large data and distributed environments. However, in the test environment,
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala:573]: `... improvements with large dataset.`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala [hunk 7]
```diff
@@ -567,6 +567,26 @@ object WholeStageCodegenExec {
...
+  // due to the small amount of data, the time to generate Java code takes up a major part of the
+  // entire runtime. So we summarize the total generation time and output it to the execution log
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala:575]: `total code generation time`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala [hunk 8]
```diff
@@ -567,6 +567,26 @@ object WholeStageCodegenExec {
...
+  // for easy analysis and view.
+  private val _generateJavaTime = new LongAccumulator
+
+  // Increase the total generation time of Java source code in nanoseconds.
+  // Visible for testing
+  def incrGenerateJavaTime(time: Long): Unit = _generateJavaTime.add(time)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala:581]: `increaseCodeGenTime`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala [hunk 9]
```diff
@@ -679,6 +700,8 @@ case class WholeStageCodegenExec(child: SparkPlan)(val codegenStageId: Int)
       new CodeAndComment(CodeFormatter.stripExtraNewLines(source), ctx.getPlaceHolderToComments()))
 
     logDebug(s"\n${CodeFormatter.format(cleanedSource)}")
+    val duration = System.nanoTime() - startTime
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala:703]: shall we do it before the debug? Otherwise we may include `CodeFormatter.format`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala (hunk 3 above)
Comment (beliefer) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala:571]: OK.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala (hunk 4 above)
Comment (beliefer) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala:42]: OK.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala (hunk 5 above)
Comment (beliefer) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala:572]: Become enlightened at once. Thank you.

Comment (beliefer) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala:572]: OK.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala (hunk 6 above)
Comment (beliefer) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala:573]: OK.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala (hunk 7 above)
Comment (beliefer) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala:575]: OK.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala (hunk 8 above)
Comment (beliefer) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala:581]: OK.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala (hunk 9 above)
Comment (beliefer) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala:703]: OK.

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by beliefer** [comment]
@cloud-fan Thanks for your help
